dashboard/modules/reporter/profile_manager.py

```diff
@@ -49,33 +49,46 @@ def _format_failed_pyspy_command(cmd, stdout, stderr) -> str:
 # If we can sudo, always try that. Otherwise, py-spy will only work if the user has
 # root privileges or has configured setuid on the py-spy script.
 async def _can_passwordless_sudo() -> bool:
-    process = await asyncio.create_subprocess_shell(
-        "sudo -n true",
+    process = await asyncio.create_subprocess_exec(
+        "sudo", "-n", "true",
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
-        shell=True,
     )
     _, _ = await process.communicate()
     return process.returncode == 0
 
 
+async def _get_pyspy_path() -> str:
+    process = await asyncio.create_subprocess_exec(
+        "which", "py-spy",
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    stdout, _ = await process.communicate()
+    if process.returncode != 0:
+        return ""
+    return stdout.decode().strip()
+
+
 class CpuProfilingManager:
     def __init__(self, profile_dir_path: str):
         self.profile_dir_path = Path(profile_dir_path)
         self.profile_dir_path.mkdir(exist_ok=True)
 
     async def trace_dump(self, pid: int, native: bool = False) -> (bool, str):
-        cmd = f"py-spy dump -p {pid}"
+        cmd = ["py-spy", "dump", "-p" , str(pid)]
         # We
         if sys.platform == "linux" and native:
-            cmd += " --native"
+            cmd.append("--native")
         if await _can_passwordless_sudo():
-            cmd = "sudo -n " + cmd.replace("py-spy", "$(which py-spy)")
-        process = await asyncio.create_subprocess_shell(
-            cmd,
+            pyspy_path = await _get_pyspy_path()
+            if pyspy_path != "":
+                cmd[0] = pyspy_path
+                cmd = ["sudo", "-n"] + cmd
```

Comment:
Why set pyspy's path dynamically here vs in the original cmd construction?

I assume its trying to preserving existing behavior? Any reason simplifying isn't a good idea?

Reply:
Ooh, I'll switch to shutil.which and always use that.